software/examples/mqtt/example-configuration.txt
# Change XXYYZZ to the UID of your DC Brick

setup:
	publish '{"mode": "drive_coast"}' to tinkerforge/request/dc_brick/XXYYZZ/set_drive_mode

	publish '{"frequency": 10000}' to tinkerforge/request/dc_brick/XXYYZZ/set_pwm_frequency # Use PWM frequency of 10 kHz

	publish '{"acceleration": 4096}' to tinkerforge/request/dc_brick/XXYYZZ/set_acceleration # Slow acceleration (12.5 %/s)

	publish '{"velocity": 32767}' to tinkerforge/request/dc_brick/XXYYZZ/set_velocity # Full speed forward (100 %)

	publish '' to tinkerforge/request/dc_brick/XXYYZZ/enable # Enable motor power

cleanup:
	# If you are done, run this to clean up
	# Stop motor before disabling motor power
	publish '{"acceleration": 16384}' to tinkerforge/request/dc_brick/XXYYZZ/set_acceleration # Fast decceleration (50 %/s) for stopping

	publish '{"velocity": 0}' to tinkerforge/request/dc_brick/XXYYZZ/set_velocity # Request motor stop

	wait for 2s # Wait for motor to actually stop: velocity (100 %) / decceleration (50 %/s) = 2 s

	publish '' to tinkerforge/request/dc_brick/XXYYZZ/disable # Disable motor power


software/examples/mathematica/ExampleConfiguration.nb.txt
Needs["NETLink`"]
LoadNETAssembly["Tinkerforge",NotebookDirectory[]<>"../../.."]

host="localhost"
port=4223
uid="XXYYZZ"(*Change XXYYZZ to the UID of your DC Brick*)

(*Create IPConnection and device object*)
ipcon=NETNew["Tinkerforge.IPConnection"]
dc=NETNew["Tinkerforge.BrickDC",uid,ipcon]
ipcon@Connect[host,port]

dc@SetDriveMode[Tinkerforge`BrickDC`DRIVEUMODEUDRIVEUCOAST]
dc@SetPWMFrequency[10000](*Use PWM frequency of 10 kHz*)
dc@SetAcceleration[4096](*Slow acceleration (12.5 %/s)*)
dc@SetVelocity[32767](*Full speed forward (100 %)*)
dc@Enable[](*Enable motor power*)

Input["Click OK to exit"]

(*Stop motor before disabling motor power*)
dc@SetAcceleration[16384](*Fast decceleration (50 %/s) for stopping*)
dc@SetVelocity[0](*Request motor stop*)
Pause[2](*Wait for motor to actually stop: velocity (100 %) / decceleration (50 %/s) = 2 s*)
dc@Disable[](*Disable motor power*)

ipcon@Disconnect[]
ReleaseNETObject[dc]
ReleaseNETObject[ipcon]


software/changelog
2011-10-06: 1.0.0 (c009c89)
- Initial version

2012-02-15: 1.1.0 (f33ee6a)
- Add support for tick types

2012-03-05: 1.1.1 (ce4afa6)
- Fix bug with timeouts on Windows (through bricklib)

2012-03-11: 1.1.2 (a0a10b7)
- Fix bug with ADC calibration readout (through bricklib)

2012-06-30: 1.1.3 (89d1ed1)
- Add get_chip_temperature and reset function

2012-07-02: 1.1.4 (4b14a66)
- Add USB hotplug support (through bricklib)

2012-09-20: 1.1.5 (9370625)
- Set interrupt pins to default (otherwise they can interfere with WIFI)

2012-10-26: 1.1.6 (66ae792)
- Change Bricklet Context size from 250 to 256 (through bricklib)

2013-01-17: 2.0.0 (c6e2870)
- Support for Protocol 2.0

2013-04-11: 2.0.1 (0eb1fdb)
- Add watchdog timer (restart after 16s of "deadlock")

2013-06-27: 2.0.2 (d3f0801)
- Use tick_task to detect USB connect/disconnect instead of interrupt

2014-07-07: 2.0.3 (63b83bf)
- Remove USB hotplug support (through bricklib)

2014-11-25: 2.3.0 (030fbc0)
- Add support for new SPI protocol

2016-04-04: 2.3.1 (e45ed60)
- Make setting velocity to 0 work while motor is disabled

2016-08-29: 2.3.2 (af79780)
- Make sure to use all chunks when writing Bricklet plugin (through bricklib)

2017-05-11: 2.3.3 (6f39bee)
- Add support for co-mcu Bricklets

2017-06-28: 2.3.4 (d0b36fce)
- Fix incompatibility between co-mcu support and LED Strip Bricklet (through bricklib)

2017-12-13: 2.3.5 (3b204424)
- Make sure Brick does not restart on suspend/resume (through bricklib)
- Refactor initial enumeration handling, fix double-enumeration bugs (through bricklib)

2018-01-11: 2.3.6 (783e4fdc)
- Fix message duplication bug (through bricklib)

2018-04-27: 2.3.7 (2b54410b)
- Fix potential deadlock between usb stack and comcu enumeration (through bricklib)

2018-07-31: 2.3.8 (f24db991)
- Add support for Isolator Bricklet (through bricklib)
- Improve Co-MCU Bricklet error handling (through bricklib)

2019-02-27: 2.3.9 (68bd0de)
- Changes through bricklib:
  - Fix Bricklet port miss-identification
  - Increase SPI stack timeout (for RED Brick)
  - Fix bugs in set/get_spitfp_* functions


software/src/config.h
/* dc-brick
 * Copyright (C) 2010-2012 Olaf Lüke <olaf@tinkerforge.com>
 *
 * config.h: DC-Brick specific configuration
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 */

#ifndef CONFIG_H
#define CONFIG_H

#include "communication.h"
#include "bricklib/drivers/board/sam3s/SAM3S.h"
#include "dc.h"

#define BRICK_FIRMWARE_VERSION_MAJOR 2
#define BRICK_FIRMWARE_VERSION_MINOR 3
#define BRICK_FIRMWARE_VERSION_REVISION 9

#define BRICK_HARDWARE_VERSION_MAJOR 1
#define BRICK_HARDWARE_VERSION_MINOR 1
#define BRICK_HARDWARE_VERSION_REVISION 0

#define BRICK_DEVICE_IDENTIFIER 11

// ************** DEBUG SETTINGS **************
#define DEBUG_SPI_STACK 1
#define DEBUG_I2C_EEPROM 0
#define DEBUG_STARTUP 1
#define DEBUG_BRICKLET 1
#define DEBUG_DC 1
//#define PROFILING
//#define PROFILING_TIME 100 // After how many seconds profiling is printed

#define DISABLE_JTAG_ON_STARTUP
//#define LOGGING_SERIAL
//#define LOGGING_LEVEL LOGGING_DEBUG
#define LOGGING_LEVEL LOGGING_NONE


// ************** BRICK SETTINGS **************

// Frequencies
#define BOARD_MCK      64000000 // Frequency of brick
#define BOARD_MAINOSC  16000000 // Frequency of oscillator
#define BOARD_ADC_FREQ 16000000 // Frequency of ADC
#define BOARD_OSC_EXTERNAL      // Use external oscillator

#define BRICK_HAS_CO_MCU_SUPPORT

// UART for console output (printf)
#define PIN_CONSOLE_RXD  {1 << 21, PIOA, ID_PIOA, PIO_PERIPH_A, PIO_DEFAULT}
#define PIN_CONSOLE_TXD  {1 << 22, PIOA, ID_PIOA, PIO_PERIPH_A, PIO_DEFAULT}

#define CONSOLE_BAUDRATE 115200
#define CONSOLE_USART    USART1
#define CONSOLE_ID       ID_USART1
#define CONSOLE_PINS     {PIN_CONSOLE_RXD, PIN_CONSOLE_TXD}

#define PINS_UART        {PIN_CONSOLE_RXD, PIN_CONSOLE_TXD}

// TWI
// TWI version
#define TWI_V3XX

// TWI stack definitions (for reading of eeproms from Bricks in stack)
#define TWI_STACK           TWI1
#define PIN_TWI_TWD_STACK   {1 << 4, PIOB, ID_PIOB, PIO_PERIPH_A, PIO_DEFAULT}
#define PIN_TWI_TWCK_STACK  {1 << 5, PIOB, ID_PIOB, PIO_PERIPH_A, PIO_DEFAULT}
#define PINS_TWI_STACK      PIN_TWI_TWD_STACK, PIN_TWI_TWCK_STACK

// TWI bricklet definitions (for bricklets, spi select and brick specific
//                           functions)
#define TWI_BRICKLET           TWI0
#define PIN_TWI_TWD_BRICKLET   {1 << 3, PIOA, ID_PIOA, PIO_PERIPH_A, PIO_DEFAULT}
#define PIN_TWI_TWCK_BRICKLET  {1 << 4, PIOA, ID_PIOA, PIO_PERIPH_A, PIO_DEFAULT}
#define PINS_TWI_BRICKLET      PIN_TWI_TWD_BRICKLET, PIN_TWI_TWCK_BRICKLET

// USB
// USB VBUS monitoring pin for USB plug and play
#define PIN_USB_DETECT  {1 << 26, PIOA, ID_PIOA, PIO_INPUT, PIO_DEFAULT}
// USB product descriptor (name of brick)
#define PRODUCT_DESCRIPTOR { \
	USBStringDescriptor_LENGTH(8), \
    USBGenericDescriptor_STRING, \
    USBStringDescriptor_UNICODE('D'), \
    USBStringDescriptor_UNICODE('C'), \
    USBStringDescriptor_UNICODE(' '), \
    USBStringDescriptor_UNICODE('B'), \
    USBStringDescriptor_UNICODE('r'), \
    USBStringDescriptor_UNICODE('i'), \
    USBStringDescriptor_UNICODE('c'), \
    USBStringDescriptor_UNICODE('k') \
}


// SPI
#define PIN_SPI_MISO        {1 << 12, PIOA, ID_PIOA, PIO_PERIPH_A, PIO_DEFAULT}
#define PIN_SPI_MOSI        {1 << 13, PIOA, ID_PIOA, PIO_PERIPH_A, PIO_DEFAULT}
#define PIN_SPI_SPCK        {1 << 14, PIOA, ID_PIOA, PIO_PERIPH_A, PIO_DEFAULT}

#define PINS_SPI            PIN_SPI_MISO, PIN_SPI_MOSI, PIN_SPI_SPCK

#define PIN_SPI_SELECT_SLAVE  {1 << 11, PIOA, ID_PIOA, PIO_PERIPH_A, PIO_DEFAULT}

// LED
#define PIN_LED_STD_BLUE    {1 << 9, PIOA, ID_PIOA, PIO_OUTPUT_1, PIO_DEFAULT}
#define PIN_LED_STD_RED     {1 << 10, PIOA, ID_PIOA, PIO_OUTPUT_1, PIO_DEFAULT}
#define PINS_STD_LED        PIN_LED_STD_BLUE, PIN_LED_STD_RED
#define PINS_LED            PINS_STD_LED

#define LED_STD_BLUE        0
#define LED_STD_RED         1


// Brick Detect
// Set low by master
// TODO: Change name
#define PIN_DETECT        {1 << 6, PIOB, ID_PIOB, PIO_OUTPUT_0, PIO_DEFAULT}



// ************** INTERRUPT PRIORITIES ***********
#define PRIORITY_EEPROM_MASTER_TWI0  1
#define PRIORITY_EEPROM_SLAVE_TWI1   1
#define PRIORITY_STACK_SLAVE_SPI     5
#define PRIORITY_PROFILING_TC0       0

// ************** BRICKLET SETTINGS **************

// Number of bricklet ports
#define BRICKLET_NUM 2

// BRICKLET A
#define BRICKLET_A_ADDRESS 84
#define BRICKLET_A_ADC_CHANNEL 4

#define BRICKLET_A_PIN_1_AD   {1 << 0, PIOB, ID_PIOB, PIO_INPUT, PIO_DEFAULT}
#define BRICKLET_A_PIN_2_DA   {1 << 13, PIOB, ID_PIOB, PIO_INPUT, PIO_DEFAULT}
#define BRICKLET_A_PIN_3_PWM  {1 << 15, PIOA, ID_PIOA, PIO_INPUT, PIO_DEFAULT}
#define BRICKLET_A_PIN_4_IO   {1 << 0, PIOA, ID_PIOA, PIO_INPUT, PIO_DEFAULT}
#define BRICKLET_A_PIN_SELECT {0, 0, 0, 0, 0}

// BRICKLET B
#define BRICKLET_B_ADDRESS 80
#define BRICKLET_B_ADC_CHANNEL 5

#define BRICKLET_B_PIN_1_AD   {1 << 1, PIOB, ID_PIOB, PIO_INPUT, PIO_DEFAULT}
#define BRICKLET_B_PIN_2_DA   {1 << 14, PIOB, ID_PIOB, PIO_INPUT, PIO_DEFAULT}
#define BRICKLET_B_PIN_3_PWM  {1 << 16, PIOA, ID_PIOA, PIO_INPUT, PIO_DEFAULT}
#define BRICKLET_B_PIN_4_IO   {1 << 1, PIOA, ID_PIOA, PIO_INPUT, PIO_DEFAULT}
#define BRICKLET_B_PIN_SELECT {0, 0, 0, 0, 0}

// ************** DC DEBUGGING **************
#if(DEBUG_DC)
#define logdcd(str, ...) do{logd("dc: " str, ##__VA_ARGS__);}while(0)
#define logdci(str, ...) do{logi("dc: " str, ##__VA_ARGS__);}while(0)
#define logdcw(str, ...) do{logw("dc: " str, ##__VA_ARGS__);}while(0)
#define logdce(str, ...) do{loge("dc: " str, ##__VA_ARGS__);}while(0)
#define logdcf(str, ...) do{logf("dc: " str, ##__VA_ARGS__);}while(0)
#else
#define logdcd(str, ...) {}
#define logdci(str, ...) {}
#define logdcw(str, ...) {}
#define logdce(str, ...) {}
#define logdcf(str, ...) {}
#endif
#endif

// *************** MC33926 PINS ***************

#define PIN_INPUT1       {1 << 23, PIOA, ID_PIOA, PIO_PERIPH_B, PIO_DEFAULT}
#define PIN_INPUT2       {1 << 24, PIOA, ID_PIOA, PIO_PERIPH_B, PIO_DEFAULT}
#define PIN_SLEW_RATE    {1 << 28, PIOA, ID_PIOA, PIO_OUTPUT_1, PIO_DEFAULT}
#define PIN_FEEDBACK     {1 << 3, PIOB, ID_PIOB, PIO_INPUT, PIO_DEFAULT}
#define PIN_ENABLE       {1 << 30, PIOA, ID_PIOA, PIO_OUTPUT_0, PIO_DEFAULT}
#define PIN_INPUT_INV    {1 << 29, PIOA, ID_PIOA, PIO_INPUT, PIO_DEFAULT}
#define PIN_STATUS_FLAG  {1 << 2, PIOB, ID_PIOB, PIO_INPUT, PIO_PULLUP}
#define PIN_DISABLE1     {1 << 17, PIOA, ID_PIOA, PIO_OUTPUT_0, PIO_DEFAULT}
#define PIN_DISABLE2     {1 << 18, PIOA, ID_PIOA, PIO_OUTPUT_1, PIO_DEFAULT}

#define PINS_DC PIN_INPUT1, PIN_INPUT2, PIN_SLEW_RATE, PIN_FEEDBACK, \
                PIN_ENABLE, PIN_INPUT_INV, PIN_STATUS_FLAG, \
                PIN_DISABLE1, PIN_DISABLE2

#define INPUT1_CHANNEL 0
#define INPUT2_CHANNEL 1
#define DISABLE1_CHANNEL 3


// ************** POWER MANAGEMENT **************
#define VOLTAGE_MAX_VALUE 4095

#define VOLTAGE_STACK_CHANNEL 3
#define VOLTAGE_STACK_MULTIPLIER 11
#define VOLTAGE_STACK_REFERENCE 3300
#define VOLTAGE_STACK_PIN {1 << 20, PIOA, ID_PIOA, PIO_INPUT, PIO_DEFAULT}
#define VOLTAGE_STACK_SWITCH_PIN {1 << 27, PIOA, ID_PIOA, PIO_OUTPUT_1, PIO_DEFAULT}

#define VOLTAGE_EXTERN_CHANNEL 2
#define VOLTAGE_EXTERN_MULTIPLIER 11
#define VOLTAGE_EXTERN_REFERENCE 3300
#define VOLTAGE_EXTERN_PIN {1 << 19, PIOA, ID_PIOA, PIO_INPUT, PIO_DEFAULT}

#define CURRENT_CONSUMPTION_CHANNEL 7
#define CURRENT_CONSUMPTION_MULTIPLIER 1
#define CURRENT_CONSUMPTION_REFERENCE 6250  // mA at 3.3V
#define CURRENT_CONSUMPTION_PIN PIN_FEEDBACK


// *************** Stack interrupt pins *********
#define PIN_STACK_INT0   {1 << 2, PIOA, ID_PIOB, PIO_INPUT, PIO_DEFAULT}
#define PIN_STACK_INT1   {1 << 5, PIOA, ID_PIOA, PIO_INPUT, PIO_DEFAULT}
#define PIN_STACK_INT2   {1 << 6, PIOA, ID_PIOA, PIO_INPUT, PIO_DEFAULT}

#define PINS_STACK       PIN_STACK_INT0, PIN_STACK_INT1, PIN_STACK_INT2
